java/arcs/core/storage/ReferenceModeStore.kt

```diff
@@ -188,11 +190,27 @@ class ReferenceModeStore private constructor(
             return modelGetter() as RefModeStoreData
         }
 
-        val deferred = CompletableDeferred<RefModeStoreData>(coroutineContext[Job.Key])
-        sendQueue.enqueueBlocking(pendingIds) {
+        val deferred = CompletableDeferred<RefModeStoreData>()
+        val sendDeferred = sendQueue.enqueueBlocking(pendingIds) {
             deferred.complete(modelGetter() as RefModeStoreData)
         }
-        return deferred.await()
+        return try {
+            // Await on the execution of the queued+blocked send item first, so that if we time-out
+            // the operation in the sendQueue will get canceled. Then await the result from the
+            // deferred with our value.
+            withTimeout(BLOCKING_QUEUE_TIMEOUT_MILLIS) {
+                sendDeferred.await()
+                deferred.await()
+            }
+        } catch (e: TimeoutCancellationException) {
```

Comment:
Should the timeout be logged? It's most likely the case that something undesirable is happening, right?

Reply:
(And I see you are logging something for the timeout below)